examples/platform_channel/linux/my_application.cc

```diff
@@ -0,0 +1,262 @@
+#include <math.h>
+#include <upower.h>
+
+#include "my_application.h"
+
+#include <flutter_linux/flutter_linux.h>
+#ifdef GDK_WINDOWING_X11
+#include <gdk/gdkx.h>
+#endif
+
+#include "flutter/generated_plugin_registrant.h"
+
+struct _MyApplication {
+  GtkApplication parent_instance;
+  char** dart_entrypoint_arguments;
+
+  FlView* view;
+
+  // Connection to UPower.
+  UpClient* up_client;
+  GPtrArray* battery_devices;
+
+  // Channel for Dart code to request the battery information.
+  FlMethodChannel *battery_channel;
+
+  // Channel to send updates to Dart code about battery charging state.
+  FlEventChannel *charging_channel;
+  gchar* last_charge_event;
+  bool emit_charge_events;
+};
+
+G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)
+
+// Check the charging state and emit an event if necessary.
+static void update_charging_state(MyApplication *self) {
+  bool charging = false;
+  for (guint i = 0; i < self->battery_devices->len; i++) {
+    UpDevice* device = static_cast<UpDevice*>(g_ptr_array_index(self->battery_devices, i));
+
+    guint state;
+    g_object_get(device, "state", &state, nullptr);
+    if (state == UP_DEVICE_STATE_CHARGING || state == UP_DEVICE_STATE_FULLY_CHARGED) {
+      charging = true;
+    }
+  }
+
+  const gchar* charge_event = charging ? "charging" : "discharging";
+
+  if (self->emit_charge_events && g_strcmp0(charge_event, self->last_charge_event) != 0) {
+    g_autoptr(GError) error = nullptr;
+    g_autoptr(FlValue) value = fl_value_new_string(charge_event);
+    if (!fl_event_channel_send(self->charging_channel, value, nullptr, &error)) {
+      g_warning("Failed to send charging event: %s", error->message);
+      return;
+    }
+    g_free(self->last_charge_event);
+    self->last_charge_event = g_strdup(charge_event);
+  }
+}
+
+// Called when a UPower device changes state.
+static void up_device_state_changed_cb(MyApplication *self, GParamSpec* pspec, UpDevice* device) {
+  update_charging_state(self);
+}
+
+// Called when UPower devices are added.
+static void up_device_added_cb(MyApplication *self, UpDevice* device) {
+  // Listen for state changes from battery_devices.
+  guint kind;
+  g_object_get(device, "kind", &kind, nullptr);
+  if (kind == UP_DEVICE_KIND_BATTERY) {
+    g_ptr_array_add(self->battery_devices, g_object_ref(device));
+    g_signal_connect_swapped(device, "notify::state", G_CALLBACK(up_device_state_changed_cb), self);
+    up_device_state_changed_cb(self, nullptr, device);
+  }
+}
+
+// Called when UPower devices are removed.
+static void up_device_removed_cb(MyApplication *self, UpDevice* device) {
+  g_ptr_array_remove(self->battery_devices, device); // FIXME
+  g_signal_handlers_disconnect_matched(device, G_SIGNAL_MATCH_FUNC, 0, 0, nullptr, reinterpret_cast<GClosure*>(up_device_state_changed_cb), nullptr);
```

Comment:
Could you clang-format this file? It looks like we don't have any enforcement in this repo, but it would be good to follow our usual style.